code/addons/test/src/node/test-manager.ts

```diff
@@ -37,42 +39,69 @@ export class TestManager {
     this.vitestManager.startVitest().then(() => options.onReady?.());
   }
 
-  async restartVitest(watchMode = false) {
+  async restartVitest({ watchMode, coverage }: { watchMode: boolean; coverage: boolean }) {
     await this.vitestManager.vitest?.runningPromise;
     await this.vitestManager.closeVitest();
-    await this.vitestManager.startVitest(watchMode);
+    await this.vitestManager.startVitest({ watchMode, coverage });
   }
 
-  async handleConfigChange(payload: TestingModuleConfigChangePayload) {
-    // TODO do something with the config
-    const config = payload.config;
+  async handleConfigChange(
+    payload: TestingModuleConfigChangePayload<any, { coverage: boolean; a11y: boolean }>
+  ) {
+    if (payload.providerId !== TEST_PROVIDER_ID) {
+      return;
+    }
+    if (this.coverage !== payload.config.coverage) {
+      try {
+        this.coverage = payload.config.coverage;
+        await this.restartVitest({ watchMode: this.watchMode, coverage: this.coverage });
+      } catch (e) {
+        this.reportFatalError('Failed to change coverage mode', e);
+      }
+    }
   }
 
   async handleWatchModeRequest(payload: TestingModuleWatchModeRequestPayload) {
-    // TODO do something with the config
-    const config = payload.config;
-
     try {
       if (payload.providerId !== TEST_PROVIDER_ID) {
         return;
       }
 
       if (this.watchMode !== payload.watchMode) {
         this.watchMode = payload.watchMode;
-        await this.restartVitest(this.watchMode);
+        await this.restartVitest({ watchMode: this.watchMode, coverage: false });
       }
     } catch (e) {
       this.reportFatalError('Failed to change watch mode', e);
     }
   }
 
-  async handleRunRequest(payload: TestingModuleRunRequestPayload) {
+  async handleRunRequest(payload: TestingModuleRunRequestPayload<Config>) {
     try {
       if (payload.providerId !== TEST_PROVIDER_ID) {
         return;
       }
 
+      const allTestsRun = (payload.storyIds ?? []).length === 0;
+      if (this.coverage) {
+        // If we have coverage enabled and we're running all stories,
+        // we have to restart Vitest AND disable watch mode
+        // otherwise the coverage report will be incorrect.
+
+        // If we're only running a subset of stories, we have to temporarily disable coverage,
+        // as a coverage report for a subset of stories is not useful.
+        await this.restartVitest({
+          watchMode: allTestsRun ? false : this.watchMode,
+          coverage: allTestsRun,
+        });
+      }
+
       await this.vitestManager.runTests(payload);
+
+      if (this.coverage && !allTestsRun) {
+        // Re-enable coverage if it was temporarily disabled because of a subset of stories was run
+        await this.restartVitest({ watchMode: this.watchMode, coverage: this.coverage });
+      }
```

Comment:
I guess this is not the only case where you have to reset to the previous state. In the UI, you can abort test runs. Should an abort action also reset to the previous state and restart Vitest?

Reply:
yes, but that happens anyway. when you abort a test-run mid-way, this function still runs to completion.

Reply:
I'm pretty sure we had code for that, but it seems @JReinhold reverted that code change. I'm also interested what the insight was.

Reply:
I refactored that logic slightly here, but the logic remains the same. I've just double checked it by:

1. modify Button stories to be slow with a 2 sec delay in each story
2. Trigger full run with coverage, see correct report
3. Modify Button component so the coverage percentage is different in the next report
4. Trigger a Focused run on Button stories, and cancel it it mid-way
5. See that terminal reports
    1. A restart with coverage _disabled_
    2. Tests run
    3. (cancel)
    4. A restart with coverage _enabled_ again
6. trigger a full run
7. See that coverage has correctly changed now, indicating that coverage is indeed re-enabled, even after a canceled focused run.